system-helper/flatpak-system-helper.c

```diff
@@ -51,6 +59,222 @@ G_DEFINE_AUTOPTR_CLEANUP_FUNC (AutoPolkitAuthorizationResult, g_object_unref)
 G_DEFINE_AUTOPTR_CLEANUP_FUNC (AutoPolkitDetails, g_object_unref)
 G_DEFINE_AUTOPTR_CLEANUP_FUNC (AutoPolkitSubject, g_object_unref)
 
+typedef struct
+{
+  FlatpakSystemHelper *object;
+  GDBusMethodInvocation *invocation;
+  GCancellable *cancellable;
+
+  guint watch_id;
+  uid_t uid;
+
+  gint ref_count;
+
+  const gchar *ref;
+  const gchar *commit;
+  gchar *root_dir_name;
+  gchar *user_dir_name;
+  gchar *child_repo_path;
+
+  GUnixMountMonitor *mount_monitor;
+  gulong mount_monitor_id;
+  guint unsuccessful_mount_timeout_id;
+  GSubprocess *bindfs;
+} FlatpakSystemHelperOngoingPull;
+
+static void
+system_helper_ongoing_pull_free (FlatpakSystemHelperOngoingPull *pull)
+{
+  g_autoptr(GFile) root_dir = g_file_new_for_path (pull->root_dir_name);
+  g_autoptr(GFile) user_dir = g_file_new_for_path (pull->user_dir_name);
+
+  g_cancellable_cancel (pull->cancellable);
+
+  g_clear_object (&pull->bindfs);
+
+  flatpak_rm_rf (root_dir, NULL, NULL);
+  flatpak_rm_rf (user_dir, NULL, NULL);
+
+  g_clear_object (&pull->mount_monitor);
+
+  g_clear_pointer (&pull->root_dir_name, g_free);
+  g_clear_pointer (&pull->user_dir_name, g_free);
+  g_clear_pointer (&pull->child_repo_path, g_free);
+
+  G_LOCK (ongoing_pulls);
+  g_ptr_array_remove (ongoing_pulls, pull);
+  G_UNLOCK (ongoing_pulls);
+
+  g_slice_free (FlatpakSystemHelperOngoingPull, pull);
+}
+
+static void
+system_helper_ongoing_pull_teardown (FlatpakSystemHelperOngoingPull *pull)
+{
+  if (pull->bindfs != NULL)
+    {
+      /* We are probably here because graceful unmounting did not happen during handle_deploy.
+         Hence, we try force unmount our bindfs directory, to make sure no open FDs can write
+         to it anymore. If this fails, we kill the subprocess and try to cleanup /etc/mtab
+         mount entry. */
+      if (umount2 (pull->user_dir_name, MNT_FORCE) == -1)
+        {
+          g_autoptr(GError) local_error = NULL;
+          glnx_throw_errno_prefix (&local_error, "Failed to unmount bindfs dir %s", pull->user_dir_name);
+          g_warning ("%s; killing subprocess", local_error->message);
+          g_clear_error (&local_error);
```

Comment:
This seems a contrived way to format the warning... why the GError detour ?

Reply:
`umount2` sets the error as `errno`. libglnx has nice API to convert errno to GError. We flag it as a warning because below we are going to kill the subprocess if we can't umount. and proceed to cleanup.